index.js

```diff
@@ -1630,7 +1634,14 @@ question.decode = function (buf, offset) {
   q.type = types.toString(buf.readUInt16BE(offset))
   offset += 2
 
-  q.class = classes.toString(buf.readUInt16BE(offset))
+  const classValue = buf.readUInt16BE(offset)
+  if (classValue >= QU_MASK) {
+    q.qu = true
+    q.class = classes.toString(classValue - QU_MASK)
+  } else {
+    q.class = classes.toString(classValue)
+  }
+
   offset += 2
 
   const qu = !!(q.class & QU_MASK)
```

Comment:
I think this section is still problematic (and there doesn't seem to be a test for decoding a question with the QU bit set) - if the bit is set, then we end up attempting a bitwise AND operation on the q.class string value (in line 1648 just below), which wipes out the string value (the result is 0).

Thank you very much for this effort - would you mind also adding a test for the decode function?

Reply:
Certainly, this change was quite some time ago. I recall this PR request occurred in 2018 or 2019. This weekend, I will take a look and make some adjustments.